crates/aptos-jwk-consensus/src/epoch_manager.rs

```diff
@@ -0,0 +1,225 @@
+                    self.on_new_epoch(reconfig_notification).await
+                },
+                event = self.jwk_updated_events.select_next_some() => {
+                    self.process_onchain_event(event)
+                },
+                (peer, rpc_request) = network_receivers.rpc_rx.select_next_some() => {
+                    self.process_rpc_request(peer, rpc_request)
+                }
+            };
+
+            if let Err(e) = handle_result {
+                error!("{}", e);
+            }
+        }
+    }
+
+    async fn await_reconfig_notification(&mut self) {
+        let reconfig_notification = self
+            .reconfig_events
+            .next()
+            .await
+            .expect("Reconfig sender dropped, unable to start new epoch");
+        self.start_new_epoch(reconfig_notification.on_chain_configs)
+            .await;
+    }
+
+    async fn start_new_epoch(&mut self, payload: OnChainConfigPayload<P>) {
+        let validator_set: ValidatorSet = payload
+            .get()
+            .expect("failed to get ValidatorSet from payload");
+
+        let epoch_state = Arc::new(EpochState {
+            epoch: payload.epoch(),
+            verifier: (&validator_set).into(),
+        });
+        self.epoch_state = Some(epoch_state.clone());
+        info!("[JWK] start_new_epoch: new_epoch={}", epoch_state.epoch);
+
+        let features = payload.get::<Features>().unwrap_or_default();
+
+        if features.is_enabled(FeatureFlag::JWK_CONSENSUS) {
+            let onchain_oidc_provider_set = payload.get::<SupportedOIDCProviders>().ok();
+            let onchain_observed_jwks = payload.get::<ObservedJWKs>().ok();
+            info!("[JWK] JWK manager init, epoch={}", epoch_state.epoch);
+            let network_sender = NetworkSender::new(
+                self.my_addr,
+                self.network_sender.clone(),
+                self.self_sender.clone(),
+            );
+            let rb = ReliableBroadcast::new(
+                epoch_state.verifier.get_ordered_account_addresses(),
+                Arc::new(network_sender),
+                ExponentialBackoff::from_millis(5),
+                aptos_time_service::TimeService::real(),
+                Duration::from_millis(1000),
+                BoundedExecutor::new(8, tokio::runtime::Handle::current()),
+            );
+            let qc_update_producer = RealCertifiedUpdateProducer::new(rb);
+
+            let jwk_consensus_manager = JWKManager::new(
+                self.consensus_key.clone(),
+                self.my_addr,
+                epoch_state.clone(),
+                Arc::new(qc_update_producer),
+                self.vtxn_pool.clone(),
+            );
+
+            let (jwk_event_tx, jwk_event_rx) = aptos_channel::new(QueueStyle::KLAST, 1, None);
+            self.jwk_updated_event_txs = Some(jwk_event_tx);
+            let (jwk_rpc_msg_tx, jwk_rpc_msg_rx) = aptos_channel::new(QueueStyle::FIFO, 100, None);
```

Comment:
Since it is per-issuer, this channel should also has size 1?

Reply:
Just checking, 100 is fine as well

Reply:
this channel is for epoch manager to forward network message to jwk manager, so not yet related to per-issuer logic?

Reply:
Is this the channel to forward peer request? Per issuer should have only 1 request per peer?

Reply:
yes it is to forward peer request. But there can be multiple concurrent requests e.g. peer 1 requesting google.com, peer 2 requesting facebook.com.

Reply:
Yes I see. you are right.